Allow CudaTensors as indices

colesbury opened the pull request:
Allow indexFill, indexCopy, and index to accept CudaTensors as indices
instead of LongTensors. We already just do a copy from LongTensor to
CudaTensor in the cutorch implementations, so this changes doesn't
result in any additional loss in precision.

dominikgrewe commented on torch/generic/Tensor.c:
```diff
@@ -383,31 +383,38 @@ static int torch_Tensor_(indexSelect)(lua_State *L)
   else
   {
-    luaL_error(L,"Tensor, number, LongTensor | Tensor, Tensor, number, LongTensor expected");
+    luaL_error(L,"Tensor, number, Tensor | LongTensor expected");
```
This should also mention the 4-argument option. How about something like this:

```lua
luaL_error(L, "[Tensor,] Tensor, number, Tensor | LongTensor expected");
```

dominikgrewe commented:
Thanks! Would you mind extending the indexSelect etc tests with an example of using a CudaTensor for the indices?

colesbury commented:
I've updated the error message and added a test for each of indexSelect, indexCopy, and indexFill.

I originally tried to extend the tests to run with both CudaTensor and LongTensor for indices, but it was tough to get that to work with compareFloatAndCuda[Tensor]. Anyways, I think this covers all the code paths, since the LongTensor and CudaTensor mostly overlap.

dominikgrewe commented:
Yes, I think that's fine. Thanks a lot.

hughperkins commented:
Nice!

hughperkins commented:
Ported (mostly) to cltorch https://github.com/hughperkins/cltorch/commit/3fa2a0976827714ec3971bb4210485ede137f3ab
